File: zalo-miniapp/src/services/content.service.ts
import { appConfig } from "@/config/app.config";
import { apiClient, buildQuery, mockDelay, type Paged } from "@/services/api";
import { mockArticles } from "@/mocks/news.mock";
import { mockRadioBulletins } from "@/mocks/radio.mock";
import { mockVideos } from "@/mocks/video.mock";
import { mockGovContacts } from "@/mocks/directory.mock";
import type { Article, ArticleType, GovContact, RadioBulletin, VideoItem } from "@/types";

/**
 * Đường dẫn phát tệp video do xã tự tải lên.
 * Tệp video CMS đăng lên là công khai (isPrivate=false) nên GET /files/:id trả
 * thẳng nội dung, không cần token và không cần link ký sẵn.
 */
export function hostedVideoUrl(fileId: string): string {
  return `${appConfig.api.baseUrl}/files/${encodeURIComponent(fileId)}`;
}

/**
 * Bóc mã video từ mọi dạng link YouTube cán bộ có thể dán vào CMS:
 * watch?v=, youtu.be/, /embed/, /shorts/, /live/. Trả null nếu không nhận ra —
 * khi đó màn chi tiết hiện nút mở ngoài thay vì nhúng khung trắng.
 */
export function youtubeId(url: string | undefined): string | null {
  if (!url) return null;
  const patterns = [
    /[?&]v=([A-Za-z0-9_-]{11})/,
    /youtu\.be\/([A-Za-z0-9_-]{11})/,
    /\/embed\/([A-Za-z0-9_-]{11})/,
    /\/shorts\/([A-Za-z0-9_-]{11})/,
    /\/live\/([A-Za-z0-9_-]{11})/,
  ];
  for (const re of patterns) {
    const m = url.match(re);
    if (m) return m[1];
  }
  return null;
}

/**
 * Nội dung công khai của xã — tin tức, video tuyên truyền, bản tin truyền thanh.
 * Nguồn: nhóm endpoint /content/public/* của backend (KHÔNG cần token).
 *
 * Cờ appConfig.api.useMocks rẽ nhánh ngay tại đây để demo offline vẫn chạy;
 * các màn hình chỉ gọi service, không biết dữ liệu đến từ đâu.
 */

/** Số bản ghi lấy về mỗi màn danh sách — đủ cho quy mô nội dung cấp xã */
const LIST_LIMIT = 50;

/** Màu bìa dự phòng khi CMS chưa chọn màu cho bài viết/video */
const FALLBACK_COVER = "var(--slate)";

/** Bản ghi bài viết như backend trả về (content.schema.ts) */
interface ApiArticle {
  _id: string;
  type: string;
  title: string;
  category?: string;
  excerpt?: string;
  /** Chỉ có ở endpoint chi tiết — danh sách dùng projection '-content' */
  content?: string;
  coverColor?: string;
  publishedAt?: string;
  views?: number;
}

interface ApiVideo {
  _id: string;
  title: string;
  topic?: string;
  duration?: string;
  views?: number;
  publishedAt?: string;
  coverColor?: string;
  description?: string;
  /** 'youtube' | 'hosted' — cán bộ chọn khi đăng ở CMS Web Quản trị */
  source?: string;
  youtubeUrl?: string;
  videoFileId?: string;
}

interface ApiRadio {
  _id: string;
  title: string;
  category?: string;
  date?: string;
  durationSeconds?: number;
  plays?: number;
}

const ARTICLE_TYPES: ArticleType[] = ["news", "event", "notice"];

function toArticleType(value: string): ArticleType {
  return ARTICLE_TYPES.includes(value as ArticleType) ? (value as ArticleType) : "news";
}

function toArticle(raw: ApiArticle): Article {
  return {
    id: raw._id,
    type: toArticleType(raw.type),
    title: raw.title,
    category: raw.category ?? "",
    excerpt: raw.excerpt ?? "",
    content: raw.content ?? "",
    coverColor: raw.coverColor || FALLBACK_COVER,
    publishedAt: raw.publishedAt ?? "",
    views: raw.views ?? 0,
  };
}

function toVideo(raw: ApiVideo): VideoItem {
  return {
    id: raw._id,
    title: raw.title,
    topic: raw.topic ?? "",
    duration: raw.duration ?? "",
    views: raw.views ?? 0,
    publishedAt: raw.publishedAt ?? "",
    coverColor: raw.coverColor || FALLBACK_COVER,
    description: raw.description ?? "",
    /* Bản ghi cũ chưa có trường này thì coi như YouTube — đó là giá trị mặc
       định của schema backend, và cũng là nhánh không cần tệp đính kèm. */
    source: raw.source === "hosted" ? "hosted" : "youtube",
    youtubeUrl: raw.youtubeUrl,
    videoFileId: raw.videoFileId,
  };
}

function toBulletin(raw: ApiRadio): RadioBulletin {
  return {
    id: raw._id,
    title: raw.title,
    category: raw.category ?? "",
    date: raw.date ?? "",
    durationSeconds: raw.durationSeconds ?? 0,
    plays: raw.plays ?? 0,
  };
}

/** Bản ghi danh bạ backend trả về */
interface ApiGovContact {
  name: string;
  title: string;
  department: string;
  phone: string;
  group: string;
  order: number;
}

export const contentService = {
  /**
   * Danh bạ chính quyền — GET /catalogs/public/directory, không cần token.
   * Backend đã sắp xếp theo nhóm rồi tới thứ tự hiển thị.
   */
  async listGovContacts(): Promise<GovContact[]> {
    if (appConfig.api.useMocks) {
      await mockDelay();
      return mockGovContacts;
    }
    const res = await apiClient.get<{ items: ApiGovContact[] }>("/catalogs/public/directory");
    return res.items.map((raw) => ({
      name: raw.name,
      title: raw.title,
      department: raw.department,
      phone: raw.phone,
      group: raw.group as GovContact["group"],
    }));
  },

  /** Bài viết đã phát hành, lọc theo loại (news | event | notice) */
  async listArticles(type?: ArticleType): Promise<Article[]> {
    if (appConfig.api.useMocks) {
      await mockDelay();
      return type ? mockArticles.filter((a) => a.type === type) : mockArticles;
    }
    const query = buildQuery({ type, page: 1, limit: LIST_LIMIT });
    const res = await apiClient.get<Paged<ApiArticle>>(`/content/public/articles${query}`);
    return res.items.map(toArticle);
  },

  /** Chi tiết bài viết — mỗi lần mở backend tăng lượt xem */
  async getArticle(id: string): Promise<Article> {
    if (appConfig.api.useMocks) {
      await mockDelay();
      const found = mockArticles.find((a) => a.id === id);
      if (!found) throw new Error("Không tìm thấy bài viết");
      return found;
    }
    return toArticle(await apiClient.get<ApiArticle>(`/content/public/articles/${encodeURIComponent(id)}`));
  },

  /** Video tuyên truyền đã phát hành */
  async listVideos(): Promise<VideoItem[]> {
    if (appConfig.api.useMocks) {
      await mockDelay();
      return mockVideos;
    }
    const res = await apiClient.get<Paged<ApiVideo>>(
      `/content/public/videos${buildQuery({ page: 1, limit: LIST_LIMIT })}`,
    );
    return res.items.map(toVideo);
  },

  /** Bản tin truyền thanh đã phát hành */
  async listRadio(): Promise<RadioBulletin[]> {
    if (appConfig.api.useMocks) {
      await mockDelay();
      return mockRadioBulletins;
    }
    const res = await apiClient.get<Paged<ApiRadio>>(
      `/content/public/radio${buildQuery({ page: 1, limit: LIST_LIMIT })}`,
    );
    return res.items.map(toBulletin);
  },
};

/** Danh sách chuyên mục xuất hiện trong dữ liệu, giữ nguyên thứ tự gặp đầu tiên */
export function distinctBy<T>(items: T[], pick: (item: T) => string): string[] {
  const seen: string[] = [];
  for (const item of items) {
    const value = pick(item);
    if (value && !seen.includes(value)) seen.push(value);
  }
  return seen;
}

File: mobile/lib/services/content_service.dart
import '../config/app_config.dart';
import '../mocks/news_mock.dart';
import '../mocks/radio_mock.dart';
import '../mocks/video_mock.dart';
import '../models/models.dart';
import 'api_client.dart';

/// Số bản ghi tải mỗi lần cho các danh sách nội dung công khai
/// (backend giới hạn tối đa 50 bản ghi/lần với nhóm /public).
const int kContentPageSize = 50;

/// Nội dung công khai cho app công dân — tin tức, video, truyền thanh.
///
/// Nhóm `/content/public/*` KHÔNG cần token nên đọc được cả khi chưa định danh.
/// Kết quả được nhớ tạm trong bộ nhớ để màn chi tiết dựng "tin/video liên quan"
/// mà không phải gọi lại backend (backend chưa có endpoint chi tiết cho video).
class ContentService {
  ContentService({ApiClient? api}) : _api = api ?? ApiClient.instance;

  final ApiClient _api;

  final Map<ArticleType, List<Article>> _articles = {};
  List<VideoItem>? _videos;
  List<RadioBulletin>? _radio;

  // --------------------------------------------------------------- Bài viết

  /// Bài viết đã phát hành theo loại (Tin tức / Sự kiện / Thông báo)
  Future<List<Article>> articles(ArticleType type, {bool refresh = false}) async {
    final cached = _articles[type];
    if (!refresh && cached != null) return cached;

    late final List<Article> items;
    if (AppConfig.useMocks) {
      await Future<void>.delayed(AppConfig.mockDelay);
      items = mockArticles.where((a) => a.type == type).toList();
    } else {
      final res = await _api.getJson(
        '/content/public/articles',
        query: {'type': type.apiValue, 'page': 1, 'limit': kContentPageSize},
        auth: false,
      );
      items = asItems(res).map(Article.fromJson).toList();
    }
    _articles[type] = items;
    return items;
  }

  /// Chi tiết bài viết — backend tự tăng lượt xem mỗi lần mở.
  /// Danh sách dùng projection `-content` nên PHẢI gọi endpoint này để có nội dung.
  Future<Article> articleDetail(String id) async {
    if (AppConfig.useMocks) {
      await Future<void>.delayed(AppConfig.mockDelay);
      final matches = mockArticles.where((a) => a.id == id);
      if (matches.isEmpty) throw const ApiException('Không tìm thấy bài viết');
      return matches.first;
    }
    final res = await _api.getJson('/content/public/articles/$id', auth: false);
    return Article.fromJson(res);
  }

  /// Bài viết cùng loại để dựng khối "Tin liên quan"
  Future<List<Article>> relatedArticles(Article article, int count) async {
    final list = await articles(article.type);
    return list.where((a) => a.id != article.id).take(count).toList();
  }

  /// Bài viết mới nhất cho khối "Tin tức mới" ở Trang chủ (không lọc theo loại)
  Future<List<Article>> latestArticles(int count) async {
    if (AppConfig.useMocks) {
      await Future<void>.delayed(AppConfig.mockDelay);
      return mockArticles.take(count).toList();
    }
    final res = await _api.getJson(
      '/content/public/articles',
      query: {'page': 1, 'limit': count},
      auth: false,
    );
    return asItems(res).map(Article.fromJson).toList();
  }

  // ------------------------------------------------------------------ Video

  Future<List<VideoItem>> videos({bool refresh = false}) async {
    final cached = _videos;
    if (!refresh && cached != null) return cached;

    late final List<VideoItem> items;
    if (AppConfig.useMocks) {
      await Future<void>.delayed(AppConfig.mockDelay);
      items = List.of(mockVideos);
    } else {
      final res = await _api.getJson(
        '/content/public/videos',
        query: {'page': 1, 'limit': kContentPageSize},
        auth: false,
      );
      items = asItems(res).map(VideoItem.fromJson).toList();
    }
    _videos = items;
    return items;
  }

  /// Chi tiết video — backend chưa có endpoint riêng nên lấy từ danh sách đã tải
  Future<VideoItem> videoDetail(String id) async {
    final list = await videos();
    final matches = list.where((v) => v.id == id);
    if (matches.isEmpty) throw const ApiException('Không tìm thấy video');
    return matches.first;
  }

  /// Video cùng chuyên mục để dựng khối "Video liên quan"
  Future<List<VideoItem>> relatedVideos(VideoItem video, int count) async {
    final list = await videos();
    return list.where((v) => v.topic == video.topic && v.id != video.id).take(count).toList();
  }

  // ------------------------------------------------------------ Truyền thanh

  Future<List<RadioBulletin>> radio({bool refresh = false}) async {
    final cached = _radio;
    if (!refresh && cached != null) return cached;

    late final List<RadioBulletin> items;
    if (AppConfig.useMocks) {
      await Future<void>.delayed(AppConfig.mockDelay);
      items = List.of(mockRadioBulletins);
    } else {
      final res = await _api.getJson(
        '/content/public/radio',
        query: {'page': 1, 'limit': kContentPageSize},
        auth: false,
      );
      items = asItems(res).map(RadioBulletin.fromJson).toList();
    }
    _radio = items;
    return items;
  }
}

/// Danh sách chuyên mục rút ra từ dữ liệu đã tải — thay cho hằng số trong lib/mocks
/// để hàng chip lọc luôn khớp nội dung backend đang phát hành.
List<String> distinctLabels(Iterable<String> values) {
  final seen = <String>[];
  for (final v in values) {
    final label = v.trim();
    if (label.isNotEmpty && !seen.contains(label)) seen.add(label);
  }
  return seen;
}
